File: internal/cred/keychain_linux.go
//go:build linux

package cred

import (
	"context"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
)

// credPath returns the path to ~/.claude/.credentials.json.
func credPath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		// %v (not %w) for the inner os error: we expose ErrClaudeTokenNotFound
		// as the sentinel so callers classify this as auth-missing, not as a
		// wrapped os error. Wrapping with %w would make errors.Is match against
		// internal os/syscall errors, which callers should not depend on.
		return "", fmt.Errorf("%w: cannot resolve home directory ($HOME unset): %v", ErrClaudeTokenNotFound, err)
	}
	return filepath.Join(home, ".claude", ".credentials.json"), nil
}

// ReadClaudeCredential returns the full credential blob from
// ~/.claude/.credentials.json (the file `claude /login` writes on Linux).
// ctx is accepted for signature parity; not used by the current implementation.
//
// File permissions are the Claude Code CLI's responsibility (it created the
// file); aistat is a read-only observer and does not police them.
func ReadClaudeCredential(ctx context.Context) (Credential, error) {
	path, err := credPath()
	if err != nil {
		return Credential{}, err
	}
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return Credential{}, ErrClaudeTokenNotFound
		}
		return Credential{}, fmt.Errorf("reading claude credentials: %w", err)
	}
	return parseClaudeCredFull(data)
}

// ReadClaudeToken returns the OAuth access token from
// ~/.claude/.credentials.json (the file `claude /login` writes on Linux).
// ctx is accepted for signature parity; not used by the current implementation.
//
// File permissions are the Claude Code CLI's responsibility (it created the
// file); aistat is a read-only observer and does not police them.
func ReadClaudeToken(ctx context.Context) (string, error) {
	c, err := ReadClaudeCredential(ctx)
	return c.AccessToken, err
}

// WriteClaudeLiveBlob atomically overwrites ~/.claude/.credentials.json with
// the given raw credential blob. The parent directory (~/.claude) is created
// with mode 0700 if absent. The file is written with mode 0600 via a
// temporary file in the same directory followed by os.Rename, so a partial
// write is never observable by ReadClaudeCredential.
func WriteClaudeLiveBlob(ctx context.Context, rawBlob []byte) error {
	path, err := credPath()
	if err != nil {
		return fmt.Errorf("cannot resolve credential path: %w", err)
	}
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return fmt.Errorf("creating credential directory: %w", err)
	}

	tmp, err := os.CreateTemp(dir, ".credentials-*.json")
	if err != nil {
		return fmt.Errorf("creating temporary credential file: %w", err)
	}
	tmpPath := tmp.Name()
	committed := false
	defer func() {
		if !committed {
			os.Remove(tmpPath)
		}
	}()

	if err := tmp.Chmod(0o600); err != nil {
		tmp.Close()
		return fmt.Errorf("setting credential file mode: %w", err)
	}
	if _, err := tmp.Write(rawBlob); err != nil {
		tmp.Close()
		return fmt.Errorf("writing credential file: %w", err)
	}
	if err := tmp.Close(); err != nil {
		return fmt.Errorf("closing credential file: %w", err)
	}

	if err := os.Rename(tmpPath, path); err != nil {
		return fmt.Errorf("installing credential file: %w", err)
	}
	committed = true
	return nil
}

File: src/cred/keychain_linux.rs
#![cfg(target_os = "linux")]

use super::{claude::parse_claude_cred, Credential, CredError};
use std::path::PathBuf;

fn cred_path() -> Result<PathBuf, CredError> {
    let home = dirs::home_dir().ok_or_else(|| {
        CredError::Other("cannot resolve home directory".into())
    })?;
    Ok(home.join(".claude").join(".credentials.json"))
}

pub fn read_claude_credential() -> Result<Credential, CredError> {
    let path = cred_path()?;
    let data = std::fs::read(&path).map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            CredError::ClaudeNotFound
        } else {
            CredError::Other(format!("reading Claude credentials: {e}"))
        }
    })?;
    parse_claude_cred(&data)
}

pub fn write_claude_live_blob(raw_blob: &[u8]) -> Result<(), CredError> {
    let path = cred_path()?;
    let dir = path.parent().unwrap();
    std::fs::create_dir_all(dir)
        .map_err(|e| CredError::Other(format!("creating Claude credentials directory: {e}")))?;

    let tmp_path = dir.join(format!(".credentials-{}.json", std::process::id()));
    {
        use std::os::unix::fs::PermissionsExt;
        std::fs::write(&tmp_path, raw_blob)
            .map_err(|e| CredError::Other(format!("writing Claude credentials file: {e}")))?;
        std::fs::set_permissions(&tmp_path, std::fs::Permissions::from_mode(0o600))
            .map_err(|e| CredError::Other(format!("setting Claude credentials file mode: {e}")))?;
    }
    std::fs::rename(&tmp_path, &path)
        .map_err(|e| CredError::Other(format!("installing Claude credentials file: {e}")))?;
    Ok(())
}
